File: domain/src/main/java/home/oleg/placenearme/interactors/EvaluateDistance.java
package home.oleg.placenearme.interactors;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import home.oleg.placenearme.models.DetailedVenue;
import home.oleg.placenearme.models.LatLng;
import home.oleg.placenearme.models.UserLocation;
import home.oleg.placenearme.repositories.DistanceRepository;
import home.oleg.placenearme.repositories.UserLocationRepository;
import io.reactivex.Single;

/**
 * Created by Oleg Sheliakin on 27.10.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public class EvaluateDistance {

    private final UserLocationRepository locationRepository;
    private final DistanceRepository distanceRepository;

    public EvaluateDistance(UserLocationRepository locationRepository, DistanceRepository distanceRepository) {
        this.locationRepository = locationRepository;
        this.distanceRepository = distanceRepository;
    }

    public Single<List<DetailedVenue>> evaluateDistance(Iterable<DetailedVenue> venues) {
        return locationRepository.getLocation().map(userLocation -> {
            List<DetailedVenue> detailedVenues = new ArrayList<>();
            for (DetailedVenue venue : venues) {
                LatLng to = new LatLng(venue.getLocation().getLat(), venue.getLocation().getLng());
                LatLng from = new LatLng(userLocation.getLat(), userLocation.getLng());
                double distance = distanceRepository.evaluate(from, to);
                venue.setDistance(distance);
                detailedVenues.add(venue);
            }

            Collections.sort(detailedVenues, (venue, t1) -> Double.compare(venue.getDistance(), t1.getDistance()));
            return detailedVenues;
        });
    }

    public Single<DetailedVenue> evaluateDistance(DetailedVenue venue) {
        return locationRepository.getLocation().map(userLocation -> {
            LatLng to = new LatLng(venue.getLocation().getLat(), venue.getLocation().getLng());
            LatLng from = new LatLng(userLocation.getLat(), userLocation.getLng());
            double distance = distanceRepository.evaluate(from, to);
            venue.setDistance(distance);
            return venue;
        });
    }
}

File: domain/src/main/java/home/oleg/placenearme/interactors/EvaluateDistance.kt
package home.oleg.placenearme.interactors

import home.oleg.placenearme.models.DetailedVenue
import home.oleg.placenearme.models.LatLng
import home.oleg.placenearme.models.Venue
import home.oleg.placenearme.repositories.DistanceRepository
import home.oleg.placenearme.repositories.UserLocationRepository
import io.reactivex.Single
import java.util.*

/**
 * Created by Oleg Sheliakin on 27.10.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
class EvaluateDistance(
        private val locationRepository: UserLocationRepository,
        private val distanceRepository: DistanceRepository) {

    fun evaluateDistance(venues: List<DetailedVenue>): Single<List<DetailedVenue>> {
        return locationRepository.location.map { userLocation ->
            val detailedVenues = ArrayList<DetailedVenue>()
            for (venue in venues) {
                val to = LatLng(venue.location.lat, venue.location.lng)
                val from = LatLng(userLocation.lat, userLocation.lng)
                val distance = distanceRepository.evaluate(from, to)
                detailedVenues.add(venue.copy(distance = distance))
            }
            detailedVenues.sortedBy { it.distance }
        }
    }

    fun evaluateDistanceVenue(venues: List<Venue>): Single<List<Venue>> {
        return locationRepository.location.map { userLocation ->
            val detailedVenues = ArrayList<Venue>()
            for (venue in venues) {
                val to = LatLng(venue.location.lat, venue.location.lng)
                val from = LatLng(userLocation.lat, userLocation.lng)
                val distance = distanceRepository.evaluate(from, to)
                detailedVenues.add(venue.copy(distance = distance))
            }
            detailedVenues.sortedBy { it.distance }
        }
    }

    fun evaluateDistance(venue: DetailedVenue): Single<DetailedVenue> {
        return locationRepository.location.map { userLocation ->
            val to = LatLng(venue.location.lat, venue.location.lng)
            val from = LatLng(userLocation.lat, userLocation.lng)
            val distance = distanceRepository.evaluate(from, to)
            venue.copy(distance = distance)
        }
    }
}
